core/DataAccess/Model.php

```diff
@@ -77,14 +88,122 @@ public function getInvalidatedArchiveIdsSafeToDelete($archiveTable)
+                        . " AND date2 = '" . $period->getDateEnd()->getDatetime() . "')";
+                }
+            }
+            $sql .= ")";
+        }
+
+        if ($segment) {
+            $nameCondition = "name LIKE '" . Rules::getDoneFlagArchiveContainsAllPlugins($segment) . "%'";
+        } else {
+            $nameCondition = "name LIKE 'done%'";
+        }
+
+        $sql .= " AND $nameCondition";
+
+        $archivesToInvalidate = Db::fetchAll($sql);
+        $idArchives = array_column($archivesToInvalidate, 'idarchive');
+
+        // update each archive as invalidated
+        if (!empty($idArchives)) {
+            $idArchives = array_map('intval', $idArchives);
+
+            $sql = "UPDATE `$archiveTable` SET `value` = " . ArchiveWriter::DONE_INVALIDATED . " WHERE idarchive IN ("
+                . implode(',', $idArchives) . ") AND $nameCondition";
+
+            Db::query($sql);
+        }
+
+        $doneFlag = Rules::getDoneFlagArchiveContainsAllPlugins($segment ?: new Segment('', []));
+
+        // we add every archive we need to invalidate + the archives that do not already exist to archive_invalidations.
+        // except for archives that are DONE_IN_PROGRESS.
+        $archivesToCreateInvalidationRowsFor = [];
+        $inProgressArchives = [];
+        foreach ($archivesToInvalidate as $row) {
+            if ($row['name'] != $doneFlag) { // only look at done flags that equal the one we are explicitly adding
+                continue;
+            }
+
+            $archivesToCreateInvalidationRowsFor[$row['idsite']][$row['period']][$row['date1']][$row['date2']] = $row['idarchive'];
+        }
+
+        $now = Date::now()->getDatetime();
+
+        $dummyArchives = [];
+        foreach ($idSites as $idSite) {
+            foreach ($allPeriodsToInvalidate as $period) {
+                if ($period->getLabel() == 'range'
+                    && !$forceInvalidateNonexistantRanges
+                ) {
+                    continue; // range
+                }
+
+                $date1 = $period->getDateStart()->toString();
+                $date2 = $period->getDateEnd()->toString();
+                $idArchive = $archivesToCreateInvalidationRowsFor[$idSite][$period->getId()][$date1][$date2] ?? null;
+
+                $dummyArchives[] = [
+                    'idarchive' => $idArchive,
+                    'name' => $doneFlag,
+                    'idsite' => $idSite,
+                    'date1' => $period->getDateStart()->getDatetime(),
+                    'date2' => $period->getDateEnd()->getDatetime(),
+                    'period' => $period->getId(),
+                    'ts_invalidated' => $now,
+                ];
+            }
+        }
+
+        $fields = ['idarchive', 'name', 'idsite', 'date1', 'date2', 'period', 'ts_invalidated'];
+        Db\BatchInsert::tableInsertBatch('archive_invalidations', $fields, $dummyArchives);
```

Comment:
fyi @diosmosis seems we need to pass the table with a prefixed name here... got this error:

```
  Try #1: LOAD DATA INFILE : SQLSTATE[42S02]: Base table or view not found: 1146 Table 'piwik3.archive_invalidations' doesn't exist[42],
  Try #2: LOAD DATA LOCAL INFILE : SQLSTATE[42000]: Syntax error or access violation: 1148 The used command is not allowed with this MySQL version[42000]
ERROR [2020-05-06 04:20:56] 5905   Uncaught exception: libs/Zend/Db/Statement/Pdo.php(234): SQLSTATE[HY093]: Invalid parameter number: parameter was not defined
SQLSTATE[HY093]: Invalid parameter number: parameter was not defined
#0 libs/Zend/Db/Statement.php(300): Zend_Db_Statement_Pdo->_execute(Array)
#1 libs/Zend/Db/Adapter/Abstract.php(479): Zend_Db_Statement->execute(Array)
#2 libs/Zend/Db/Adapter/Pdo/Abstract.php(238): Zend_Db_Adapter_Abstract->query('INSERT IGNORE I...', Array)
#3 core/Db/Adapter/Pdo/Mysql.php(309): Zend_Db_Adapter_Pdo_Abstract->query('INSERT IGNORE I...', Array)
#4 core/Db.php(276): Piwik\Db\Adapter\Pdo\Mysql->query('INSERT IGNORE I...', Array)
#5 core/Db/BatchInsert.php(41): Piwik\Db::query('INSERT IGNORE I...', Array)
#6 core/Db/BatchInsert.php(133): Piwik\Db\BatchInsert::tableInsertBatchIterate('archive_invalid...', Array, Array)
#7 core/DataAccess/Model.php(202): Piwik\Db\BatchInsert::tableInsertBatch('archive_invalid...', Array, Array)
#8 core/Archive/ArchiveInvalidator.php(433): Piwik\DataAccess\Model->updateArchiveAsInvalidated('piwik_archive_n...', Array, Array, NULL, false)
#9 core/Archive/ArchiveInvalidator.php(285): Piwik\Archive\ArchiveInvalidator->markArchivesInvalidated(Array, Array, NULL, true, false)
#10 plugins/CoreAdminHome/API.php(163): Piwik\Archive\ArchiveInvalidator->markArchivesAsInvalidated(Array, Array, 'day', NULL, false, false)
#11 core/CronArchive.php(955): Piwik\Plugins\CoreAdminHome\API->invalidateArchivedReports(Array, Array, 'day')
#12 core/CronArchive.php(875): Piwik\CronArchive->invalidateRecentDate('yesterday')
#13 core/CronArchive.php(341): Piwik\CronArchive->invalidateArchivedReportsForSitesThatNeedToBeArchivedAgain()
#14 core/CronArchive.php(258): Piwik\CronArchive->run()
```

Reply:
just the prefix locally in my test but then still got the bound parameter wrong error

Reply:
This is the content of `$dummyArchives` which actually looked good:

```php
array (
  0 => 
  array (
    'idarchive' => '22',
    'name' => 'done',
    'idsite' => 1,
    'date1' => '2020-05-04 00:00:00',
    'date2' => '2020-05-04 00:00:00',
    'period' => 1,
    'ts_invalidated' => '2020-05-06 04:25:23',
  ),
  1 => 
  array (
    'idarchive' => '21',
    'name' => 'done',
    'idsite' => 1,
    'date1' => '2020-05-04 00:00:00',
    'date2' => '2020-05-10 00:00:00',
    'period' => 2,
    'ts_invalidated' => '2020-05-06 04:25:23',
  ),
  2 => 
  array (
    'idarchive' => '19',
    'name' => 'done',
    'idsite' => 1,
    'date1' => '2020-05-01 00:00:00',
    'date2' => '2020-05-31 00:00:00',
    'period' => 3,
    'ts_invalidated' => '2020-05-06 04:25:23',
  ),
)
```

resulting in `Uncaught exception: parameter number: parameter was not defined`

It generated this query:
```sql
INSERT IGNORE INTO piwik_archive_invalidations
					  (idarchive,name,idsite,date1,date2,period,ts_invalidated)
					  VALUES (?,?,?,?,?,?,?)
```

The bind value looks good so not quite sure what causes it
![image](https://user-images.githubusercontent.com/273120/81138620-c62ed800-8fb6-11ea-956e-f4c0637f46d4.png)
![image](https://user-images.githubusercontent.com/273120/81138686-fbd3c100-8fb6-11ea-86c1-65eac64fd9f3.png)
Not quite understanding it... Maybe you can reproduce when eg disabling `[General] enable_load_data_infile = 1`. Otherwise must be something with my install maybe.

Reply:
Hmm, weird, will look into it.

Reply:
Fixed this in another PR: https://github.com/matomo-org/matomo/pull/15919